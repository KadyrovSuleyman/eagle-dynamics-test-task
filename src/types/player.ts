export interface Player {
  id: string,
  status: string,
  name: string,
  avatar: string,
  description?: string,
}
