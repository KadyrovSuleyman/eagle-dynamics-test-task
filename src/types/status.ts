const STATUS = {
  OFFLINE: 'offline',
  ONLINE: 'online',
};

export default Object.freeze(STATUS);
export type StatusT = typeof STATUS;
