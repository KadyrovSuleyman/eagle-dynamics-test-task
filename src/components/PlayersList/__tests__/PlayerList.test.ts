import Vuex, { Store } from 'vuex';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import { Player } from '@/types/player';
import mockedList from '../../../store/__mocks__/list';
import PlayersListVue from '../PlayersList.vue';

jest.mock('../state.ts');
jest.mock('../handlers.ts');

const actions = {
  select: jest.fn(),
};

let store: Store<{ list: Player[] }>;
beforeEach(() => {
  store = new Vuex.Store({
    state: {
      // list: mockedList.map((item) => ({
      //   ...item,
      //   selected: false,
      // })),
      list: mockedList,
    },
    mutations: {
      deleteLast: () => { store.state.list.pop(); },
      modifyFirst: () => {
        store.state.list[0].id = 'id';
        store.state.list[0].name = 'name';
        store.state.list[0].avatar = 'avatar';
        store.state.list[0].description = 'description';
        store.state.list[0].status = 'status';
      },
    },
    actions,
  });
});

const localVue = createLocalVue();
localVue.use(Vuex);
let wrapper: Wrapper<Vue>;
beforeEach(() => {
  wrapper = mount(PlayersListVue, { localVue, store });
});

it('renders filled list', () => {
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

it('renders empty list', async () => {
  store.state.list = [];
  await wrapper.vm.$nextTick();
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

describe('watchs outer store', () => {
  it('modify some players data', async () => {
    expect(wrapper.findAll('.playersList-playerItem').length).toBe(mockedList.length);
    expect(wrapper.findAll('.playersList-playerItem').at(0)).toMatchSnapshot();

    store.commit('modifyFirst');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.playersList-playerItem').at(0)).toMatchSnapshot();
  });

  it('modify players list itself', async () => {
    const { length } = mockedList;
    expect(wrapper.findAll('.playersList-playerItem').length).toBe(length);
    expect(wrapper.element.outerHTML).toMatchSnapshot();

    store.commit('deleteLast');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.playersList-playerItem').length).toBe(length - 1);
    expect(wrapper.element.outerHTML).toMatchSnapshot();
  });
});

it('triggers onclick func', async () => {
  expect(actions.select).toBeCalledTimes(0);

  await wrapper.findAll('.playersList-playerItem').at(0).trigger('click');
  expect(actions.select).toBeCalledTimes(1);
  expect(actions.select.mock.calls[0][1]).toBe(store.state.list[0].id);

  await wrapper.findAll('.playersList-playerItem').at(1).trigger('click');
  expect(actions.select).toBeCalledTimes(2);
  expect(actions.select.mock.calls[1][1]).toBe(store.state.list[1].id);
});
