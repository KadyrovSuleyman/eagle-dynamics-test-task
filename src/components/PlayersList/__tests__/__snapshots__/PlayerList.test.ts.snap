// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty list 1`] = `
<div class="app-playersListFallback">
  There are no players yet
</div>
`;

exports[`renders filled list 1`] = `
<div class="app-playersList">
  <div class="playersList-playerItem playersList-playerItem__status_offline">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="🛦 Flying ℃omradE 🛦" class="playerItem-img"></div>
    <div class="playerItem-name">
      🛦 Flying ℃omradE 🛦
    </div>
    <div class="playerItem-status">
      offline
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="𝕯esperate 𝕭ℓⱭⅆe 𝕽unner" class="playerItem-img"></div>
    <div class="playerItem-name">
      𝕯esperate 𝕭ℓⱭⅆe 𝕽unner
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face3.jpg" alt="Fluffy" class="playerItem-img"></div>
    <div class="playerItem-name">
      Fluffy
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="Appel82" class="playerItem-img"></div>
    <div class="playerItem-name">
      Appel82
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="No. 428" class="playerItem-img"></div>
    <div class="playerItem-name">
      No. 428
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="&quot;Matic&quot;" class="playerItem-img"></div>
    <div class="playerItem-name">
      "Matic"
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="“Maverick”" class="playerItem-img"></div>
    <div class="playerItem-name">
      “Maverick”
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="braveheart" class="playerItem-img"></div>
    <div class="playerItem-name">
      braveheart
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="-=~0verthehor!zon~=-" class="playerItem-img"></div>
    <div class="playerItem-name">
      -=~0verthehor!zon~=-
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face3.jpg" alt="Justdoit" class="playerItem-img"></div>
    <div class="playerItem-name">
      Justdoit
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
</div>
`;

exports[`watchs outer store modify players list itself 1`] = `
<div class="app-playersList">
  <div class="playersList-playerItem playersList-playerItem__status_status">
    <div class="playerItem-avatar"><img src="avatar" alt="name" class="playerItem-img"></div>
    <div class="playerItem-name">
      name
    </div>
    <div class="playerItem-status">
      status
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="𝕯esperate 𝕭ℓⱭⅆe 𝕽unner" class="playerItem-img"></div>
    <div class="playerItem-name">
      𝕯esperate 𝕭ℓⱭⅆe 𝕽unner
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face3.jpg" alt="Fluffy" class="playerItem-img"></div>
    <div class="playerItem-name">
      Fluffy
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="Appel82" class="playerItem-img"></div>
    <div class="playerItem-name">
      Appel82
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="No. 428" class="playerItem-img"></div>
    <div class="playerItem-name">
      No. 428
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="&quot;Matic&quot;" class="playerItem-img"></div>
    <div class="playerItem-name">
      "Matic"
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="“Maverick”" class="playerItem-img"></div>
    <div class="playerItem-name">
      “Maverick”
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="braveheart" class="playerItem-img"></div>
    <div class="playerItem-name">
      braveheart
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="-=~0verthehor!zon~=-" class="playerItem-img"></div>
    <div class="playerItem-name">
      -=~0verthehor!zon~=-
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face3.jpg" alt="Justdoit" class="playerItem-img"></div>
    <div class="playerItem-name">
      Justdoit
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
</div>
`;

exports[`watchs outer store modify players list itself 2`] = `
<div class="app-playersList">
  <div class="playersList-playerItem playersList-playerItem__status_status">
    <div class="playerItem-avatar"><img src="avatar" alt="name" class="playerItem-img"></div>
    <div class="playerItem-name">
      name
    </div>
    <div class="playerItem-status">
      status
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="𝕯esperate 𝕭ℓⱭⅆe 𝕽unner" class="playerItem-img"></div>
    <div class="playerItem-name">
      𝕯esperate 𝕭ℓⱭⅆe 𝕽unner
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face3.jpg" alt="Fluffy" class="playerItem-img"></div>
    <div class="playerItem-name">
      Fluffy
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="Appel82" class="playerItem-img"></div>
    <div class="playerItem-name">
      Appel82
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face2.jpg" alt="No. 428" class="playerItem-img"></div>
    <div class="playerItem-name">
      No. 428
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="&quot;Matic&quot;" class="playerItem-img"></div>
    <div class="playerItem-name">
      "Matic"
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="“Maverick”" class="playerItem-img"></div>
    <div class="playerItem-name">
      “Maverick”
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face4.jpg" alt="braveheart" class="playerItem-img"></div>
    <div class="playerItem-name">
      braveheart
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
  <div class="playersList-playerItem playersList-playerItem__status_online">
    <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="-=~0verthehor!zon~=-" class="playerItem-img"></div>
    <div class="playerItem-name">
      -=~0verthehor!zon~=-
    </div>
    <div class="playerItem-status">
      online
    </div>
  </div>
</div>
`;

exports[`watchs outer store modify some players data 1`] = `
<div class="playersList-playerItem playersList-playerItem__status_offline">
  <div class="playerItem-avatar"><img src="dev/profiles/face1.jpg" alt="🛦 Flying ℃omradE 🛦" class="playerItem-img"></div>
  <div class="playerItem-name">
    🛦 Flying ℃omradE 🛦
  </div>
  <div class="playerItem-status">
    offline
  </div>
</div>
`;

exports[`watchs outer store modify some players data 2`] = `
<div class="playersList-playerItem playersList-playerItem__status_status">
  <div class="playerItem-avatar"><img src="avatar" alt="name" class="playerItem-img"></div>
  <div class="playerItem-name">
    name
  </div>
  <div class="playerItem-status">
    status
  </div>
</div>
`;
